### DISCUSSION TITLE
Smart hints

### DISCUSSION THREAD

**Discussion by lua-rocks** [orbitalquark/textadept#89] [Q&A] [answered]
I would like to set up smart hints in textadept (based on comments like emmylua), but I have no idea how. I tried to connect the lsp module, but it does not work (I guess that I need to separately download and configure the server, but it's all very difficult, I don't know where and how). I also noticed that textadept gives smart hints for basic lua functionality based on ldoc and maybe adding ldoc-based hints to my code would be easier than lsp, but I also have no information on how to implement this. Can anyone please make a short guide for a noob? Thanks.

**Comment by orbitalquark** [comment]
https://orbitalquark.github.io/textadept/manual.html#generate-autocompletion-and-documentation-files has information on how to use LuaDoc to generate api files from your modules/code and use them with Textadept.

**Reply by lua-rocks** [reply]
@orbitalquark thanks for your answer, but I can't get this to work.
```
> luadoc -d ./api --doclet /opt/textadept/modules/lua/tadoc.lua
/usr/bin/luajit: /usr/local/share/lua/5.1/luadoc/init.lua:39: module '/opt/textadept/modules/lua/tadoc.lua' not found:No LuaRocks module found for /opt/textadept/modules/lua/tadoc.lua
```
I have installed `luadoc`, `ldoc` and `luafilesystem` for `luajit` and `lua5.3`.
The file `/opt/textadept/modules/lua/tadoc.lua` existed and I have full permissions in `/opt` directory.

**Reply by lua-rocks** [reply]
And I also don't understand this [module(s)] at the end:

`luadoc -d [output_path] --doclet _HOME/modules/lua/tadoc.lua [module(s)]`

Are they optional? Because I want to have api for **all** modules from my project.

...Anyway, the command `luadoc -d ./api --doclet /opt/textadept/modules/lua/tadoc.lua ./main.lua` also doesn't work.

**Reply by lua-rocks** [reply]
When I remove luadoc for luajit:
`sudo luarocks --lua-version=5.1 remove luadoc`

and reinstall it for lua5.3:
`sudo luarocks --lua-version=5.3 install luadoc`

I got new error:
```
> luadoc -d ./api --doclet /opt/textadept/modules/lua/tadoc.lua ./main.lua
/usr/bin/lua5.3: /usr/local/share/lua/5.3/luadoc/util.lua:12: attempt to call a nil value (global 'module')
```

**Reply by lua-rocks** [reply]
🤔

**Reply by lua-rocks** [reply]
I don't understand anything here, but it seems to me that only you can generate api-files because you have some *special* version of luadoc which is inaccessible to mere mortals.

**Reply by lua-rocks** [reply]
I hope I wrong about luadoc version and my mistake in something else, bust just in case I installed and configured luaver, so now I can try it with any version of lua. Tell me please, what version of luadoc you use, where I can get it and what version of lua it is required?

**Comment by orbitalquark** [comment, marked as answer]
Oh no, sorry for all the trouble :( There are a few things here to address.

1. I am running LuaDoc with Lua 5.1. I don't think it can be run with Lua 5.3 or greater because the `module` function was removed.
2. The documentation is wrong in that you need to supply the doclet **without** the ".lua" extension. I need to fix this.
3. On my machine, LuaDoc does not like absolute paths for doclets, so I get an error like "no file './/opt/textadept/modules/lua/tadoc.lua'". But I don't have LuaRocks installed, which may allow for absolute paths because your error message is different.

If I wanted to document all of the modules in my _USERHOME, I would run this from Textadept's home directory:

    luadoc -d /tmp --doclet modules/lua/tadoc /home/username/.textadept/modules

The *tags* and *api* files will be placed in */tmp*. Unfortunately the filepaths in the tags will be mangled (which only affects usage with the ctags module), but you can use a find & replace script or Textadept to strip the prefix.

I never ran into this because the Makefile always invokes LuaDoc using relative paths[1]. Sorry for the trouble :(

[1]: https://github.com/orbitalquark/textadept/blob/faf8a95eea2a27ea1f776c9e61882bc27af9c735/src/Makefile#L255

**Reply by lua-rocks** [reply]
@orbitalquark Don't worry, it's okay. Thank you so much! I managed it! 🎉

BTW, don't you want to switch from luadoc to [ldoc](http://stevedonovan.github.io/ldoc/manual/doc.md.html)?

**Comment by orbitalquark** [comment]
I'm pretty sure that would involve rewriting my doclet, which I would prefer not to do.

**Reply by lua-rocks** [reply]
Ok, this is reasonable.

**Reply by lua-rocks** [reply]
@orbitalquark to my surprise, your editor is the only one in the universe which use luadoc for tooltips! Therefore, I consider it the best. Most modern editors prefer to use emmylua, which tries to be smarter than you, but in fact is terribly stupid, doesn't understand how metamethods works and displays completely ridiculous hints.

**Reply by lua-rocks** [reply]
There was another good one, called LDT, but died 😭
